SwiftCompilerSources/Sources/Optimizer/FunctionPasses/ReleaseDevirtualizer.swift

```diff
@@ -0,0 +1,140 @@
+//===--- ReleaseDevirtualizer.swift - Devirtualizes release-instructions --===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2022 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SIL
+
+/// Devirtualizes release instructions which are known to destruct the object.
+///
+/// This means, it replaces a sequence of
+///    %x = alloc_ref [stack] $X
+///      ...
+///    strong_release %x
+///    dealloc_stack_ref %x
+/// with
+///    %x = alloc_ref [stack] $X
+///      ...
+///    set_deallocating %x
+///    %d = function_ref @dealloc_of_X
+///    %a = apply %d(%x)
+///    dealloc_stack_ref %x
+///
+/// The optimization is only done for stack promoted objects because they are
+/// known to have no associated objects (which are not explicitly released
+/// in the deinit method).
+let releaseDevirtualizerPass = FunctionPass(
+  name: "release-devirtualizer", { function, context in
+    let functionInfo = context.rcIdentityAnalysis.getFunctionInfo(function)
+
+    for block in function.blocks {
+      // The last `release_value`` or `strong_release`` instruction before the
+      // deallocation.
+      var lastRelease: RefCountingInst?
+
+      for instruction in block.instructions {
+        if let release = lastRelease {
+          // We only do the optimization for stack promoted object, because for
+          // these we know that they don't have associated objects, which are
+          // _not_ released by the deinit method.
+          if let deallocStackRef = instruction as? DeallocStackRefInst {
+            devirtualizeReleaseOfObject(context, release, deallocStackRef, functionInfo)
+            lastRelease = nil
+            continue
+          }
+        }
+
+        if instruction is ReleaseValueInst || instruction is StrongReleaseInst {
+          lastRelease = instruction as? RefCountingInst
+        } else if instruction.mayReleaseOrReadRefCount {
+          lastRelease = nil
+        }
+      }
+    }
+  }
+)
+
+/// Devirtualize releases of array buffers.
+private func devirtualizeReleaseOfObject(
+  _ context: PassContext,
+  _ release: RefCountingInst,
+  _ deallocStackRef: DeallocStackRefInst,
+  _ functionInfo: RCIdentityFunctionInfo
+) {
+  let allocRefInstruction = deallocStackRef.allocRef
+  let rcRoot = functionInfo.getRCIdentityRoot(release.operands[0].value)
+  var root = release.operands[0].value
+  while let newRoot = getRCIdentityPreservingCastOperand(root) {
+    root = newRoot
+  }
+
+  precondition(rcRoot === root)
+
+  guard rcRoot === allocRefInstruction else {
```

Comment:
No need for a `guard`. You can use an `if`.
Also, no need for `===`. `Value` has a `==` operator.

Reply:
`allocRefInstruction` here is of `AllocRefInst` type, while `root` is of `Value`. For some reason `==` does not automatically cast `AllocRefInst` to `Value`, would you prefer me to cast explicitly, or keep `===` as it is?

Reply:
Also, `Value` isn't `Equatable` right now, should the former inherit from the latter? Otherwise `!=` isn't currently defined on `Value` arguments to convert this from `guard` to `if` in a nice way.